### PR TITLE
options support: overrides

### PR DESCRIPTION
**Pull request by mpenet**
As I mentioned in #30 I added an option param to coerce/conform/coerce! that allows to specify ::overrides. From there we propagate that value without using the dynamic var *overrides*. 
Along the way I fixed bugs (ex: non preserving meta on map-of) and improved perfs of some fns like `spec-is-homogeneous-set?`.

Performance match the original for common cases. I suspect there are a few cases where it would be faster, improve with usage over (run)time, and since now the internals are more "pure" it would open doors to potential optimisations (like memoization/sharing of sym->coercer return predicates), some of which could be made optional via the new arg.

The api is backward compatible, I made sure that if no option is passed we just use the dynamic var as option. Eventually we could imagine getting rid of it. I took the liberty to mark it as deprecated.

T̶h̶e̶ ̶*̶*̶o̶n̶l̶y̶*̶*̶ ̶a̶r̶e̶a̶ ̶w̶h̶e̶r̶e̶ ̶t̶h̶e̶ ̶a̶p̶i̶ ̶s̶u̶r̶f̶a̶c̶e̶ ̶c̶h̶a̶n̶g̶e̶d̶ ̶(̶I̶ ̶t̶h̶i̶n̶k̶)̶,̶ ̶i̶s̶ ̶i̶f̶/̶w̶h̶e̶n̶ ̶y̶o̶u̶ ̶c̶a̶l̶l̶ ̶p̶a̶r̶s̶e̶-̶x̶x̶x̶ ̶d̶i̶r̶e̶c̶t̶l̶y̶,̶ ̶s̶i̶n̶c̶e̶ ̶n̶o̶w̶ ̶t̶h̶e̶y̶ ̶t̶a̶k̶e̶ ̶2̶ ̶a̶r̶g̶s̶,̶ ̶t̶h̶e̶ ̶x̶ ̶(̶v̶a̶l̶u̶e̶)̶ ̶a̶n̶d̶ ̶t̶h̶e̶ ̶o̶p̶t̶i̶o̶n̶s̶ ̶(̶s̶i̶n̶c̶e̶ ̶w̶e̶ ̶n̶e̶e̶d̶ ̶t̶o̶ ̶b̶e̶ ̶a̶b̶l̶e̶ ̶t̶o̶ ̶p̶o̶t̶e̶n̶t̶i̶a̶l̶l̶y̶ ̶p̶r̶o̶p̶a̶g̶a̶t̶e̶ ̶t̶h̶e̶m̶ ̶t̶o̶ ̶i̶n̶t̶e̶r̶n̶a̶l̶ ̶`̶c̶o̶e̶r̶c̶e̶`̶ ̶c̶a̶l̶l̶s̶)̶.̶ ̶I̶ ̶t̶h̶i̶n̶k̶ ̶i̶n̶ ̶p̶r̶a̶c̶t̶i̶c̶e̶ ̶t̶h̶a̶t̶'̶s̶ ̶n̶o̶t̶ ̶s̶o̶m̶e̶t̶h̶i̶n̶g̶ ̶p̶r̶o̶b̶l̶e̶m̶a̶t̶i̶c̶ ̶f̶o̶r̶ ̶l̶i̶b̶r̶a̶r̶y̶ ̶c̶o̶n̶s̶u̶m̶e̶r̶s̶,̶ ̶w̶e̶ ̶c̶o̶u̶l̶d̶ ̶a̶l̶s̶o̶ ̶i̶m̶a̶g̶i̶n̶e̶ ̶m̶a̶k̶i̶n̶g̶ ̶t̶h̶e̶ ̶"̶l̶e̶a̶f̶"̶ ̶p̶r̶e̶d̶i̶c̶a̶t̶e̶s̶,̶ ̶e̶x̶ ̶p̶a̶r̶s̶e̶-̶l̶o̶n̶g̶ ̶&̶ ̶o̶t̶h̶e̶r̶s̶ ̶s̶u̶p̶p̶o̶r̶t̶ ̶1̶ ̶a̶r̶g̶ ̶a̶n̶d̶ ̶2̶,̶ ̶w̶h̶e̶r̶e̶ ̶a̶r̶i̶t̶y̶-̶1̶ ̶w̶o̶u̶l̶d̶ ̶d̶i̶s̶p̶a̶t̶c̶h̶ ̶o̶n̶ ̶2̶+̶n̶i̶l̶,̶ ̶t̶o̶ ̶m̶a̶k̶e̶ ̶i̶t̶ ̶1̶0̶0̶%̶ ̶c̶o̶m̶p̶a̶t̶i̶b̶l̶e̶ ̶(̶I̶ ̶m̶i̶g̶h̶t̶ ̶j̶u̶s̶t̶ ̶d̶o̶ ̶i̶t̶ ̶n̶o̶w̶)̶.̶ 
I added the extra arity like I mentioned in the previous paragraph. An alternative could be to have a small wrapper fn that wraps the simple scalar parse-xxx functions that use no options:

```clj
(defn parse-no-opts [f]
  (fn [x _] (f x)))

(defmethod sym->coercer `number? [_] (parse-no-opts parse-double))
(defmethod sym->coercer `int? [_] (parse-no-opts parse-long))
(defmethod sym->coercer `pos-int? [_] (parse-no-opts parse-long))
[...]
(defmethod sym->coercer `s/or [form] (parse-or form))
(defmethod sym->coercer `s/coll-of [form] (parse-coll-of form))
(defmethod sym->coercer `s/map-of [form] (parse-map-of form))
(defmethod sym->coercer `s/tuple [form] (parse-tuple form))
(defmethod sym->coercer `s/multi-spec [form] (parse-multi-spec form))
```
full example here, (parse-xxx untouched, dispatch handled at sym->coerce level), : https://github.com/wilkerlucio/spec-coerce/commit/04c02ac580a0463639c931ff6b9f81a25d6e93e6

I am fine with either way, maybe the parse-no-opts makes the source a bit cleaner.

### CODE REVIEW

**Comment by wilkerlucio** [comment]
Thanks, looks great! Can you update the branch? I just add config to run tests on branches here :)

**Comment by mpenet** [comment]
> Thanks, looks great! Can you update the branch? I just add config to run tests on branches here :)

Done!

**Comment by mpenet** [comment]
added a bugfix for `(sc/coerce ::nilable-string nil)` that was returning "", now returns nil

**Comment by mpenet** [comment]
also fixed coerce on s/keys missing an associative check

File: src/spec_coerce/core.cljc [hunk 1]
```diff
@@ -389,12 +449,23 @@
...
-     x)))
+                   (cond->> x
+                     (map? x)
+                     (into (empty x)
+                           (map (fn [[k v]]
+                                  (let [coercion (get overrides k k)]
+                                    [k (op coercion v opts)]))))))
+                 x)))
+
+(s/def ::foo (s/nilable string?))
```
Comment (wilkerlucio) [inline src/spec_coerce/core.cljc:463]: Can we move these examples as tests? Or maybe just remove then from here

Comment (mpenet) [reply src/spec_coerce/core.cljc]: done, I already had tests covering these. I just removed these examples

**Pull request merged**